### PR TITLE
Fix density matrix expval cython function

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Fixes bug in #5941 that showed up in Aer CI tests.

Also fixed the phase coefficient of the `Pauli` operator not being applied if it was not `1`.

### Details and comments

The cython function for density matrix expectation value only worked if the matrix memory was a C-contiguous array, which isnt always true since superop evolution vectorizes the array using Fortran ordering (col-vec). This updates the cython function to work on a flattened array assuming F ordering and adds some extra tests to check it works with both C- and F-ordered data arrays in density matrix.

### CODE REVIEW

**Review by mtreinish — COMMENTED** [review]
While we're fixing things can you add the cpp file generated by cython to the git ignore list: https://github.com/Qiskit/qiskit-terra/blob/master/.gitignore#L141-L142 (we have them in place to cover transpiler passes but it doesn't cover this new module)

**Review by mtreinish — DISMISSED** [review]
LGTM, the extra test coverage is a good addition because I think this cython module was not fully covered before which led to this issue and some other potential ones (see: #5989)

**Comment by nonhermitian** [comment]
There is still a lot of Python overhead here, e.g:

Screen Shot 2021-03-11 at 06 10 42

this can be viewed by calling `cythonize -a FILENAME.pyx`.

For the loops I would add the decorator `@cython.boundscheck(False)` to the functions.

**Comment by mtreinish** [comment]
I took care of the unnecessary python overhead in: https://github.com/Qiskit/qiskit-terra/pull/5984/commits/a4892efd15c552781d770bf25d114ae3bc7aa0b6

That being said I think there are further optimizations we can use like using the gcc builtin popcount function which will use special instructions if they're available. But I think we can optimize this in a follow up. This PR was more a bug fix because the original cython module honestly merged before it was ready and had a lot of issues and poor test coverage.

**Review by mtreinish — DISMISSED** [review]

**Review by mtreinish — APPROVED** [review]

File: qiskit/quantum_info/states/statevector.py [hunk 1]
```diff
@@ -372,17 +372,22 @@ def _expectation_value_pauli(self, pauli):
         # pylint: disable=no-name-in-module
         from .cython.exp_value import expval_pauli_no_x, expval_pauli_with_x
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply qiskit/quantum_info/states/statevector.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/5984/commits/396c86267318d2302482976cb5014c60ef2e8e03

**Pull request merged**